flink-table/flink-sql-jdbc-driver/src/main/java/org/apache/flink/table/jdbc/DriverUri.java

```diff
@@ -0,0 +1,208 @@
+        // remove last item due to a trailing slash
+        if (parts.get(parts.size() - 1).isEmpty()) {
+            parts = parts.subList(0, parts.size() - 1);
+        }
+
+        if (parts.size() > 2) {
+            throw new SQLException("Invalid path segments in URL: " + uri);
+        }
+
+        if (parts.get(0).isEmpty()) {
+            throw new SQLException("Catalog name in URL is empty: " + uri);
+        }
+
+        catalog = Optional.ofNullable(parts.get(0));
+
+        if (parts.size() > 1) {
+            if (parts.get(1).isEmpty()) {
+                throw new SQLException("Database name in URL is empty: " + uri);
+            }
+
+            database = Optional.ofNullable(parts.get(1));
+        }
+    }
+
+    private static Properties mergeDynamicProperties(URI uri, Properties driverProperties)
+            throws SQLException {
+        Map<String, String> urlProperties = parseUriParameters(uri.getQuery());
+        Map<String, String> suppliedProperties = Maps.fromProperties(driverProperties);
+
+        for (String key : urlProperties.keySet()) {
+            if (suppliedProperties.containsKey(key)) {
+                throw new SQLException(
+                        format("Connection property '%s' is both in the URL and an argument", key));
+            }
+        }
+
+        Properties result = new Properties();
+        setMapToProperties(result, urlProperties);
+        setMapToProperties(result, suppliedProperties);
+        return result;
+    }
+
+    private static void setMapToProperties(Properties properties, Map<String, String> values) {
+        for (Map.Entry<String, String> entry : values.entrySet()) {
+            properties.setProperty(entry.getKey(), entry.getValue());
+        }
+    }
+
+    private static Map<String, String> parseUriParameters(String query) throws SQLException {
+        Map<String, String> result = new HashMap<>();
+
+        if (query != null) {
+            Iterable<String> queryArgs = URL_ARG_SPLITTER.split(query);
+            for (String queryArg : queryArgs) {
+                List<String> parts = ARG_VALUE_SPLITTER.splitToList(queryArg);
+                if (parts.size() != 2) {
+                    throw new SQLException(
+                            format(
+                                    "Connection property in uri must be key=val format: '%s'",
+                                    queryArg));
+                }
+                if (result.put(parts.get(0), parts.get(1)) != null) {
+                    throw new SQLException(
+                            format(
+                                    "Connection property '%s' is in URL multiple times",
+                                    parts.get(0)));
+                }
+            }
+        }
+
+        return result;
+    }
+
+    private static URI parseDriverUrl(String url) throws SQLException {
+        if (!url.startsWith(URL_START)) {
+            throw new SQLException("Invalid Flink JDBC URL: " + url);
```

Comment:
Should we add more messages to indicate why it's invalid? same to below exception messages.